src/stamps.ts

```diff
@@ -1,28 +1,235 @@
+ *
+ * @returns Newly bought postage stamp
  */
-export function shouldReplaceStamp({ POSTAGE_STAMP }: PostageStamps = {}): boolean {
-  return Boolean(POSTAGE_STAMP)
+export async function buyNewStamp(
+  depth: number,
+  amount: string,
+  beeDebug: BeeDebug,
+  pollingFrequency = 1_000,
+): Promise<DebugPostageBatch> {
+  return new Promise(async (resolve, reject) =>
+    beeDebug
+      .createPostageBatch(amount, depth)
+      .then(async (batchId: BatchId) => {
+        waitUntilStampUsable(batchId, beeDebug, pollingFrequency)
+          .then(resolve)
+          .catch(e => {
+            logger.error('error: failed to wait until postage stamp is usable', e)
+            reject(e)
+          })
+      })
+      .catch(e => {
+        logger.error('error: failed to buy postage stamp', e)
+        reject(e)
+      }),
+  )
+}
+
+export class StampsManager {
+  private stamp?: string
+  private usableStamps?: DebugPostageBatch[]
+  public readonly enabled: boolean
+  private interval?: ReturnType<typeof setInterval>
+
+  constructor(
+    {
+      POSTAGE_STAMP,
+      POSTAGE_DEPTH,
+      POSTAGE_AMOUNT,
+      POSTAGE_USAGE_THRESHOLD,
+      POSTAGE_USAGE_MAX,
+      POSTAGE_TTL_MIN,
+      POSTAGE_REFRESH_PERIOD,
+    }: Args = {},
+    BEE_DEBUG_API_URL?: string,
+  ) {
+    this.enabled = true
+
+    if (POSTAGE_STAMP) this.stamp = POSTAGE_STAMP
+    else if (POSTAGE_DEPTH && POSTAGE_AMOUNT && BEE_DEBUG_API_URL) {
+      const refreshPeriod = Number(POSTAGE_REFRESH_PERIOD || 60_000)
+      this.start(
+        Number(POSTAGE_DEPTH),
+        POSTAGE_AMOUNT,
+        Number(POSTAGE_USAGE_THRESHOLD || '0.7'),
+        Number(POSTAGE_USAGE_MAX || '0.95'),
+        Number(POSTAGE_TTL_MIN || (refreshPeriod / 1_000) * 5),
+        refreshPeriod,
+        new BeeDebug(BEE_DEBUG_API_URL),
+      )
+    } else if (POSTAGE_DEPTH || POSTAGE_AMOUNT || BEE_DEBUG_API_URL) {
+      throw new Error('Please provide POSTAGE_DEPTH, POSTAGE_AMOUNT and BEE_DEBUG_API_URL environment variable')
+    } else this.enabled = false
+  }
+
+  /**
+   * Get postage stamp that should be replaced in a the proxy request header
+   *
+   * @return Postage stamp that should be used by the proxy
+   *
+   * @throws Error if there is no postage stamp
+   */
+  get postageStamp(): string {
+    if (this.stamp) return this.stamp
+
+    if (this.usableStamps && this.usableStamps[0]) return this.usableStamps[0].batchID
+
+    throw new Error('No postage stamp')
+  }
+
+  /**
+   * Refresh stamps from the bee node and if needed buy new stamp
+   *
+   * @param depth Postage stamps depth
+   * @param amount Postage stamps amount
+   * @param usageThreshold Threshold when new postage stamp should be bought
+   * @param maxUsage Maximal usage of the stamp to be usable by the proxy
+   * @param minTTL Minimal TTL of the stamp to be usable by the proxy
+   * @param beeDebug Connection to debug endpoint for checking/buying stamps
+   */
+  public async refreshStamps(
```

Comment:
FYI there can be race conditions for buying the stamps here if the refresh interval is too small. Semaphore would solve it, but not sure if you want to introduce that.

Reply:
Yeah so this is exactly what I wanted to avoid. More logic. I would leave it until such problem arises and in the meantime recommend refresh periods around 5 mins (or for sure at least 1 minute)